pkg/operator/configobserver/oauth/observe_tokenconfig.go

```diff
@@ -0,0 +1,150 @@
+package oauth
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog"
+
+	configlistersv1 "github.com/openshift/client-go/config/listers/config/v1"
+	"github.com/openshift/library-go/pkg/operator/configobserver"
+	"github.com/openshift/library-go/pkg/operator/events"
+)
+
+// OAuthLister lists OAuth information
+type OAuthLister interface {
+	OAuthLister() configlistersv1.OAuthLister
+}
+
+const (
+	defaultAccessTokenMaxAgeSeconds          = float64(86400) // a day
+	fieldAccessTokenMaxAgeSeconds            = "accessTokenMaxAgeSeconds"
+	fieldAccessTokenInactivityTimeoutSeconds = "accessTokenInactivityTimeoutSeconds"
+)
+
+// ObserveAccessTokenMaxAgeSeconds returns an unstructured fragment of KubeAPIServerConfig that changes the default value for access token max age,
+// if there is a valid value for it in OAuth cluster config.
+func ObserveAccessTokenMaxAgeSeconds(genericlisters configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (ret map[string]interface{}, errs []error) {
+	errs = []error{}
+	tokenConfigPath := []string{"oauthConfig", "tokenConfig"}
+	tokenMaxAgePath := append(tokenConfigPath, fieldAccessTokenMaxAgeSeconds)
+	defer func() {
+		// Prune the observed config so that it only contains access token max age field.
+		ret = configobserver.Pruned(ret, tokenMaxAgePath)
+	}()
+
+	listers, ok := genericlisters.(OAuthLister)
+	if !ok {
+		return existingConfig, append(errs, fmt.Errorf("failed to assert: given lister does not implement an OAuth lister"))
+	}
+
+	existingAccessTokenMaxAgeSeconds, _, err := unstructured.NestedFloat64(existingConfig, tokenMaxAgePath...)
+	if err != nil {
+		// If failed to read the config for access token max age field, continue execution with the
+		// hope to find the value for it in OAuth cluster config.
+		errs = append(errs, err)
+	}
+
+	oauthConfig, err := listers.OAuthLister().Get("cluster")
+	if err != nil {
+		// Failed to read OAuth cluster config.
+		if errors.IsNotFound(err) {
+			klog.Warning("oauth.config.openshift.io/cluster: not found")
+		}
+
+		if existingAccessTokenMaxAgeSeconds == 0 {
+			// As the value 0 here indicates that a read from existing config failed previously, use the default value.
+			return buildUnstructuredTokenConfig(defaultAccessTokenMaxAgeSeconds, tokenMaxAgePath), append(errs, err)
+		}
```

Comment:
just return existingConfig here and the err appended to errs. This extra logic is wrong IMO. The nested float might be missing, and the lister returns an error for whatever reason. This code here tries to be clever, but makes the situation worse.

Reply:
my understanding was that this observer must either return the default value or whatever it observed. didn't realize that an empty config can be returned. hence the clever tricks ;)

removed the code now.

Reply:
I checked other config observers in the library-go. All those observers are using default values when cluster config is not found. Shouldn't this observer follow the same?